gnome-image-installer/util/gis-unattended-config.c

```diff
@@ -409,3 +409,109 @@ gis_unattended_config_match_computer (GisUnattendedConfig *self,
 
   return GIS_UNATTENDED_COMPUTER_DOES_NOT_MATCH;
 }
+
+static GFile *
+invent_backup_name (GFile *source)
+{
+  g_autoptr(GFile) dir = g_file_get_parent (source);
+  g_autofree gchar *basename = g_file_get_basename (source);
+  g_autofree gchar *uuid = g_uuid_string_random ();
+  gchar *dot = strrchr (basename, '.');
+  g_autofree gchar *target_basename = NULL;
+
+  if (dot == NULL)
+    {
+      target_basename = g_strjoin (".", basename, uuid, NULL);
+    }
+  else
+    {
+      *dot = '\0';
+      target_basename = g_strjoin (".", basename, uuid, dot + 1, NULL);
+    }
+
+  return g_file_get_child (dir, target_basename);
+}
+
+static gboolean
+rename_existing_file_if_exists (GFile   *source,
+                                gchar  **backup_path,
+                                GError **error)
+{
+  g_autoptr(GFile) target = invent_backup_name (source);
+  g_autoptr(GError) local_error = NULL;
+
+  if (g_file_move (source, target,
+                   /* No G_FILE_COPY_OVERWRITE, so target will not be
+                    * overwritten if it already exists. Since its name includes
+                    * a fresh UUID, this is extremely unlikely.
+                    */
+                   G_FILE_COPY_NO_FALLBACK_FOR_MOVE,
+                   NULL, NULL, NULL, &local_error))
+    {
+      *backup_path = g_file_get_basename (target);
+      return TRUE;
+    }
+
+  if (g_error_matches (local_error,
+                       G_IO_ERROR,
+                       G_IO_ERROR_NOT_FOUND))
+    {
+      backup_path = NULL;
+      return TRUE;
+    }
+
+  g_propagate_error (error, g_steal_pointer (&local_error));
+  return FALSE;
+}
+
+/**
+ * gis_unattended_config_write:
+ * @target_path: where to write new unattended config. If it already exists,
+ *  the existing file will be renamed to a unique name in the same directory.
+ * @backup_path: (out): set to the basename of backup of @target_path, if it
+ *  already existed, or to %NULL if it did not.
+ *
+ * Either @vendor and @product must both be %NULL, or they must both be
+ * non-%NULL.
```

Comment:
⇒ `g_return_val_if_fail ((vendor == NULL) == (product == NULL), FALSE);`.

Reply:
This assertion is made below:

```c
  if (vendor != NULL)
    {
      g_return_val_if_fail (product != NULL, FALSE);
      // ...
    }
  else
    {
      g_return_val_if_fail (product == NULL, FALSE);
    }
```

This makes any assertion message more precise.